launchpod/src/test/java/com/google/launchpod/FormHandlerServletTest.java

```diff
@@ -0,0 +1,403 @@
+import java.util.Arrays;
+import java.util.Collection;
+import java.util.Collections;
+import java.util.List;
+
+import org.junit.Assert;
+import org.junit.Before;
+import org.junit.Rule;
+import org.junit.After;
+import org.junit.Test;
+import org.junit.rules.ExpectedException;
+import org.junit.runner.RunWith;
+import org.junit.runners.JUnit4;
+import org.mockito.InjectMocks;
+import org.mockito.Mock;
+import org.mockito.Mockito;
+import org.mockito.MockitoAnnotations;
+
+/** 
+ * Runs unit tests for the FormHandlerServlet that contains doPost(), doGet(), and xmlString() methods.
+*/
+@RunWith(JUnit4.class)
+public class FormHandlerServletTest extends Mockito {
+
+    @InjectMocks
+    private FormHandlerServlet servlet = new FormHandlerServlet();
+
+    @Mock
+    HttpServletRequest request;
+
+    @Mock
+    HttpServletResponse response;
+
+    @Rule // JUnit 4 uses Rules for testing specific messages
+    public ExpectedException thrown = ExpectedException.none();
+
+    private final LocalServiceTestHelper helper =
+    new LocalServiceTestHelper(new LocalDatastoreServiceTestConfig()
+    .setDefaultHighRepJobPolicyUnappliedJobPercentage(100));
+
+    // keys
+    private static final String USER_FEED = "UserFeed";
+    private static final String PODCAST_TITLE = "title";
+    private static final String XML_STRING = "xmlString";
+    private static final String MP3_LINK = "mp3link"; // URL to existing MP3 file
+    private static final String TIMESTAMP = "timestamp";
+    private static final String EMAIL = "email"; 
+    private static final String ID = "id";
+
+    private static final String TEST_PODCAST_TITLE = "TEST_PODCAST_TITLE";
+    private static final String TEST_EMPTY_PODCAST_TITLE = "";
+    private static final String TEST_NULL_PODCAST_TITLE = null;
+    private static final String TEST_MP3_LINK = "TEST_MP3_LINK";
+    private static final String TEST_EMPTY_MP3_LINK = "";
+    private static final String TEST_NULL_MP3_LINK = null;
+    private static final long TEST_TIMESTAMP = System.currentTimeMillis();
+    private static final String TEST_ID = "123456";
+    private static final String TEST_ID_TWO = "789012";
+    private static final String TEST_XML_STRING;
+    private static final String TEST_PUBDATE = "2020/06/26 01:32:06";
+    private static final String TEST_EMAIL = "123@abc.com";
+    private static final String TEST_INCORRECT_EMAIL = "123@cde.com";
+    private static final String TEST_EMAIL_TWO = "456@abc.com";
+    private static final String EMPTY_STRING = "";
+    private static final String BASE_URL = "https://launchpod-step18-2020.appspot.com?id=";
+
+    @Before 
+    public void setUp() {
+        MockitoAnnotations.initMocks(this);
+        helper.setUp();
+    }
+
+    @After 
+    public void tearDown() {
+      helper.tearDown();
+    }
+
+    /**
+     * Creates a test user feed entity.
+    */
+    private Entity makeEntity(String title, String mp3Link, String xmlString) {
```

Comment:
is there a method in the main servlet code that does this same thing? probably worth creating a static helper method for this and use it in both places

Reply:
the method in the main servlet creates an object from an entity, this method just makes an entity to make setting up datastore easier